### PR TITLE
gh-81163: Add support for extended color functions in ncurses 6.1

### PR DESCRIPTION
**Pull request by hpjansson**
This is a rebase of @websurfer5's #13534 against master, since that PR no longer applies cleanly and has not been updated since May. It now appears to pass the check that was failing in https://github.com/python/cpython/pull/13534#pullrequestreview-243758708.

The original description from @websurfer5's PR follows.

---

ncurses 6.1 adds extended color functions to support terminals with 256 colors (e.g. xterm-256color). The extended functions use color-pair values that are signed integers because the existing functions only use signed short integer values that are too small to support the 65,536 color-pair values needed for a 256-color terminal.

The new extended color functions are used transparently by curses.color_content(), curses.init_color(), curses.init_pair(), and curses.pair_content() when available, and the original functions are used when they aren't. The module functions validate their parameters as signed integers when using the new extended functions and as signed short integers when using the original library functions to match the underlying ncurses function parameters. The original behavior remains unchanged when Python is built with a curses library that does not contain the new extended color functions.

A new function, curses.has_extended_color_support(), indicates whether extended color support is provided by the underlying ncurses library.

This PR also fixes issue #36630.

* Issue: gh-81163

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@hpjansson

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by hpjansson** [comment]
I signed the CLA earlier today, hopefully it shows up in the system soon.

**Comment by dnastase** [comment]
Hi,

May I ask what is the status of this fix ? 
This issue affects a lot of ncurses based projects.

Thanks.

File: Doc/whatsnew/3.8.rst
Comment (ZackerySpytz) [inline Doc/whatsnew/3.8.rst]: This entry should be moved to `Doc/whatsnew/3.9.rst`.

File: Modules/_cursesmodule.c
Comment (ZackerySpytz) [inline Modules/_cursesmodule.c]: The indentation is off here (and in the next block).

File: Modules/_cursesmodule.c
Comment (ZackerySpytz) [inline Modules/_cursesmodule.c]: Please limit lines to 79 chars (https://www.python.org/dev/peps/pep-0007/#code-lay-out).

For what it's worth, I think the error messages added in this PR are too vague.

File: Modules/_cursesmodule.c
Comment (ZackerySpytz) [inline Modules/_cursesmodule.c]: There is a space missing before the "is".

File: Modules/_cursesmodule.c
Comment (ZackerySpytz) [inline Modules/_cursesmodule.c]: It seems that this hunk was copied from the generated Argument Clinic code in `Modules/clinic/_cursesmodule.c.h`. I don't think that's correct.

File: Doc/whatsnew/3.8.rst
Comment (hpjansson) [reply Doc/whatsnew/3.8.rst]: Fixed.

File: Modules/_cursesmodule.c
Comment (hpjansson) [reply Modules/_cursesmodule.c]: Fixed the formatting and made the error messages a bit more specific.

File: Modules/_cursesmodule.c
Comment (hpjansson) [reply Modules/_cursesmodule.c]: Fixed.

File: Modules/_cursesmodule.c
Comment (hpjansson) [reply Modules/_cursesmodule.c]: Fixed.

File: Modules/_cursesmodule.c
Comment (hpjansson) [reply Modules/_cursesmodule.c]: Looks like the original PR was somewhat unfinished. I removed the copied code, changed the _curses_pair_content_impl() argument type from _NCURSES_COLOR_VAL_TYPE to int, then added preconditions to ensure the passed-in value is within range (determined at compile time). This is in line with how it's done elsewhere in the patch, and it made the clinic output sane too.

**Comment by hpjansson** [comment]
Thanks for the review. I've addressed the issues brought up so far and a few more I found, and rebased it on master.

**Comment by ned-deily** [comment]
Alas, soon after the PR was updated, a large change across a number of modules relating to argument processing was merged by @serhiy-storchaka (578c395 for bpo-37999) that now causes merge conflicts for this PR. It would also be nice to merge this to 3.9 but there seem to be other merge conflicts there.  Perhaps after fixing the merge conflicts here, a 3.9 backport PR could be prepared.

**Comment by serhiy-storchaka** [comment]
Thank you @websurfer5 and @hpjansson for your PR. The code looks good to me in general, but it can be made even better.

1. The code contains repetitions. It would be better to add a purposed converter (see the documentation for `O&` at https://docs.python.org/3/c-api/arg.html#other-objects)
```c
static int
color_converter(PyObject *arg, void *ptr)
{
    ...
}

/*[python input]
class color_converter(CConverter):
    type = 'int'
    converter = 'color_converter'
[python start generated code]*/
```
and use it in Argument Clinic as
```
    color_number: color
        The number of the color to be changed (0 - COLORS).
```
And similar for color pair. And maybe also for color components.

2. I think that `ValueError` may be better exception than `OverflowError`. You can also use more appropriate error message (mentioning ranges `(0 - COLORS)` and `(1 - (COLOR_PAIRS-1))`). You can use `PyLong_AsLongAndOverflow()`.

3. It is worth to expose constants `COLORS` and `COLOR_PAIRS`.

After finishing this add your name as the co-author name of the PR in What's New. Add your and Jeffrey's names in `Misc/ACKS` if they are still not here.

**Comment by hpjansson** [comment]
Thanks for the review!

>     1. I think that `ValueError` may be better exception than `OverflowError`. You can also use more appropriate error message (mentioning ranges `(0 - COLORS)` and `(1 - (COLOR_PAIRS-1))`). You can use `PyLong_AsLongAndOverflow()`.

Done.

>     2. It is worth to expose constants `COLORS` and `COLOR_PAIRS`.

Isn't this already done in _curses_start_color_impl()? It has to be done after the call to start_color(), since it assigns the correct values to these global variables.

> After finishing this add your name as the co-author name of the PR in What's New. Add your and Jeffrey's names in `Misc/ACKS` if they are still not here.

Done.

**Comment by ned-deily** [comment]
@serhiy-storchaka WIll you be able to review soon the changes you requested?  We really should get this change merged and backported soon; ncurses 6.1 and now 6.2 are out in the field.

File: Modules/_cursesmodule.c
Comment (serhiy-storchaka) [inline Modules/_cursesmodule.c]: It is -1 if the argument is larger than LONG_MAX.

I would write:
```c
if (overflow > 0 || color_number > COLORS) {
    ...
}
else if (overflow < 0 || color_number < 0) {
    ...
}
```
Please add also tests which cover the following cases: `-1`, `COLORS+1`, `-2**31-1`, `2**31`, `-2**63-1`, `2**63`.

File: Modules/_cursesmodule.c
Comment (serhiy-storchaka) [inline Modules/_cursesmodule.c]:
```suggestion
    if (color_number == -1 && PyErr_Occurred())
```
For performance.

File: Modules/_cursesmodule.c
Comment (serhiy-storchaka) [inline Modules/_cursesmodule.c]: Why `color_pair_to_attr()` has been inlined? This change looks unrelated.

File: Modules/_cursesmodule.c
Comment (serhiy-storchaka) [inline Modules/_cursesmodule.c]: Why it was added? This change looks unrelated.

File: Modules/_cursesmodule.c
Comment (serhiy-storchaka) [inline Modules/_cursesmodule.c]: It adds 11 lines of code (8 lines to define `_CURSES_INIT_COLOR_FUNC` and `_CURSES_INIT_COLOR_FUNC_NAME` and other 3 lines to undefine them). But using `#if/#else/#endif` at the call place would add only 4 lines of code. Would not it be simpler?

The same question is about other defines used only once.

Comment (serhiy-storchaka) [reply Modules/_cursesmodule.c]: Other option is to define all macros in a single `#if/#else/#endif` construction:
```c
#if _NCURSES_EXTENDED_COLOR_FUNCS
#  define _NCURSES_COLOR_VAL_TYPE   int
#  define _COLOR_CONTENT_FUNC       extended_color_content
#  define _CURSES_INIT_COLOR_FUNC   init_extended_color
#  define _CURSES_INIT_PAIR_FUNC    init_extended_pair
#  define _CURSES_PAIR_NUMBER_FUNC  extended_pair_content
#else
...
```

Comment (hpjansson) [reply Modules/_cursesmodule.c]: I went with the latter. It seems more clear than leaving it scattered throughout the code.

**Comment by hpjansson** [comment]
@serhiy-storchaka All done.

**Comment by ned-deily** [comment]
@serhiy-storchaka, ping. This is waiting for your final review.

**Review by ned-deily — CHANGES_REQUESTED** [review]
The `has_extended_color_support()` documentation change for `Doc/library/curses.rst` that is in #13534 seems to have been missed here.  Please include a version of it.  Thanks!

**Comment by websurfer5** [comment]
@ned-deily, I assume @hpjansson has to copy the file since it is his pull request?

**Comment by hpjansson** [comment]
I have made the requested changes; please review again.

@websurfer5 Since this is your work originally, I gave you access to my fork -- feel free to change it as you like.

**Review by ned-deily — APPROVED** [review]
Thanks for adding the doc change.

I built and ran test_curses on macOS with both the system `ncurses` 5.x and with a current `ncurses` 6.2 and verified that the current `test_curses` fails with 6.2 and that, with the PR applied, `test_curses` passes with both versions of `ncurses` and that `has_extended_color_support()` returns the expected values.

**Comment by dnastase** [comment]
Nice ! thank you all for your work and efforts !

**Comment by christianbundy** [comment]
Thanks for fixing this!

**Comment by ned-deily** [comment]
@ambv and I discussed backporting this for 3.9.0 but I had forgotten that, because of other changes already in 3.10, backporting this to 3.9 is not trivial so I can't recommend trying to throw it in just prior to 3.9.0rc1.

**Pull request merged**
